stellargraph/data/explorer.py

```diff
@@ -414,30 +399,21 @@ def run(
         if weighted:
             self._check_weights_valid()
 
-            # calculate the appropriate unnormalised transition probability, given the history of
-            # the walk
-            def transition_probability(nn):
-                nn, weight = nn
-
-                if nn == previous_node:  # d_tx = 0
-                    return ip * weight
-                elif any(nn == pn for pn, _ in previous_node_neighbours):  # d_tx = 1
-                    return weight
-                else:  # d_tx = 2
-                    return iq * weight
+        weight_dtype = self.graph._edges.weights.dtype
+        cast_func = np.cast[weight_dtype]
+        ip = cast_func(1.0 / p)
+        iq = cast_func(1.0 / q)
 
-        else:
-            # without weights
-            def transition_probability(nn):
-                if nn == previous_node:  ## d_tx = 0
-                    return ip
-                elif nn in previous_node_neighbours:  # d_tx = 1
-                    return 1.0
-                else:  # d_tx = 2
-                    return iq
-
-        ip = 1.0 / p
-        iq = 1.0 / q
+        if np.isinf(ip):
+            raise ValueError(
+                f"p: value is too small. It must be possible to "
+                f"represent 1/p in {str(weight_dtype)}"
```

Comment:
FWIW, I think it's harder to read to wrap this (but maybe your IDE is doing it automatically)?

Also, we can easily give the user a little more guidance by, e.g. including the actual value, so they can understand what went wrong more directly.

Finally, `f` string `{...}` formatting is already implicitly calling `str` on the value.

E.g.:

```suggestion
                f"p: value ({p}) is too small. It must be possible to represent 1/p in {weight_dtype}, but this value overflows to infinity."
```